---
title: "RNAseq Workflow"
author: "Thomas W. Battaglia"
date: "`r Sys.Date()`"
output: hrbrmrkdn::bulma
---


### Introduction  
RNA seq is an emerging technology for deep RNA sequencing. Not only does RNA seq technology have the ability to analyze differences in gene expression between samples, but can discover new isoforms of genes and analyze SNP variation of particular genes. This tutorial will cover the basic workflow for analyzing RNA seq data on host tissue. It will not cover the workflow for analyzing RNA seq on microbial community samples. For analysis of the meta-transcriptome using RNA seq, see the workflow for Whole Genome Sequencing Metagenomics (WGS Metagenomics) in a different tutorial.

#### 1a. Setup new directory for analysis
Many files are generated during the RNA seq workflow, so it is best to keep them stored in an organized directory. The first step before running/installing any tools, is to generate a directory structure for easier organization. The rest of this tutorial assumes your working directory (```pwd```) is within the ```new_rnaseq_analysis``` folder created during the first step, or whatever you choose to name the folder.  
<br>
**input** : per sample sequencing files (.fastq)  
**genome** : directory for storing genome of interest (.fasta/.fna)  
**annotation** : directory for storing the annotation file associated with host genome (.gtf/.gff3)  
**index** : directory that will store STAR aligner's index files   
**tools** : directory to store dependencies required for analysis  
**venv** : a virtual environment of python to store all dependencies  
**logs** : a directory to store all the logs files generated from SGE  

  
```{bash, eval = F}
# Make new root directory
mkdir new_rnaseq_analysis

# Change directory into new folder
cd new_rnaseq_analysis

# Make sub-directories to store:
mkdir -p input
mkdir -p genome
mkdir -p annotation
mkdir -p index
mkdir -p tools
mkdir -p logs
```

#### 1b. Setup new virtual environment to store tools
A new virtual environment will keep all the tools that are required for analysis in its own folder so that there are no errors intorduced into the system python installation and/or break and tools that require a particular version. Most of the tools require python 2.7+. After creating and activating the virtualenv, you should see ```(venv)``` on the command line. To check what version of python you are currently using, run the command ```python --version```.  

```{bash, eval = F}
# Load python (if not set on Python 2.7)
module load python/2.7.3

# Install the virtualenv package
pip install virtualenv --user

# Make a new virtual environment folder in the 
virtualenv venv --python=python2.7

# Start the virtualenv
source venv/bin/activate

# Create a function to link tools to virtualenv
# Sourced from http://huttenhower.sph.harvard.edu/docs/anadama_workflows/install.html#id1
function link() { ln -sv $(readlink -f "$1") $(pwd)/venv/bin/; }
```
Note: If there is an error during the virtualenv creation step that states ```The executable python2.7 (from --python=python2.7) does not exist```, then the cluster environment's python version is not set for Python 2.7+, so the command ```module load python/2.7.3``` must be run first before running the ```virtualenv venv --python=python2.7``` command.


#### 1c. Download an example dataset (Optional)
If you want to use an example dataset to practice the RNAseq alignment workflow, run the commands below to place a fastq file in the ```input``` folder.  
<br>
**Citation: ** https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4124169/?tool=EBI  
The transcriptional landscape of mouse beta cells compared to human beta cells reveals notable species differences in long non-coding RNA and protein-coding gene expression. Benner C., van der Meulen T., Cacéres E., Tigyi K., Donaldson CJ., Huising MO.BMC Genomics 15620 (2014)  
```{bash, eval = F}
# Download publically available mouse RNAseq fastq files.
# http://www.ebi.ac.uk/ena/data/view/PRJNA252388
wget -P input/ ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR137/001/SRR1374921/SRR1374921.fastq.gz
wget -P input/ ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR137/002/SRR1374922/SRR1374922.fastq.gz
wget -P input/ ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR137/003/SRR1374923/SRR1374923.fastq.gz
wget -P input/ ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR137/004/SRR1374924/SRR1374924.fastq.gz

```


### 2. Installing required tools
This workflow requires many different tools. Some of these tools maybe available on your cluster environment, but to ensure the correct versioning of the tools, it is suggested to install them again inside a virtual environment. See http://docs.python-guide.org/en/latest/dev/virtualenvs/ for more information about using virtual-environments.
<br>
Tools that are required for processing are:  
a. **cutadapt**  
b. **FastQC**  
c. **Trim Galore!**  
d. **SortMeRNA**  
e. **STAR-aligner**  
f. **Subread**  
g. **MultiQC**  


#### 2a. Install cutadapt
"Cutadapt finds and removes adapter sequences, primers, poly-A tails and other types of unwanted sequence from your high-throughput sequencing reads."  
<br>
**Link:** http://cutadapt.readthedocs.io/en/stable/guide.html  
```{bash, eval = F}
pip install cutadapt
```

#### 2b. Install FastQC
"FastQC aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis."  
<br>
**Link:** http://www.bioinformatics.babraham.ac.uk/projects/fastqc/  
```{bash, eval = F}
# Download FastQC to the 'tools' folder  
wget -P tools/ http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.5.zip  

# Unzip and remove compressed file
unzip tools/fastqc_v0.11.5.zip -d tools/
rm -rf tools/fastqc_v0.11.5.zip

# Change permissions (only for fastqc tool)
chmod 777 tools/FastQC/fastqc

# Link package to virtual environment 
link tools/FastQC/fastqc

# Check if FastQC was installed properly
fastqc --version
```

#### 2c. Install Trim Galore!
"Trim Galore! is a wrapper script to automate quality and adapter trimming as well as quality control, with some added functionality to remove biased methylation positions for RRBS sequence files (for directional, non-directional (or paired-end) sequencing)."  
<br>
**Link:** http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/  

```{bash, eval = F}
# Download Trim Galore! to the 'tools' folder
wget -P tools/ http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/trim_galore_v0.4.1.zip 

# Unzip and remove compressed file
unzip tools/trim_galore_v0.4.1.zip -d tools/
rm -rf tools/trim_galore_v0.4.1.zip

# Link package to virtual environment 
link tools/trim_galore_zip/trim_galore

# Check if Trim galore! was installed properly
trim_galore --version
```

#### 2d. Install SortMeRNA
"SortMeRNA is a biological sequence analysis tool for filtering, mapping and OTU-picking NGS reads. The core algorithm is based on approximate seeds and allows for fast and sensitive analyses of nucleotide sequences. The main application of SortMeRNA is filtering rRNA from metatranscriptomic data."  
<br>
**Link:** http://bioinfo.lifl.fr/RNA/sortmerna/  
```{bash, eval = F}
# Download SortMeRNA to the 'tools' folder
wget -P tools/ https://github.com/biocore/sortmerna/archive/2.1.tar.gz 

# Unzip and remove compressed file
tar -zxvf tools/2.1.tar.gz -C tools/
rm -rf tools/2.1.tar.gz

# Compile scripts
cd tools/sortmerna-2.1/ && bash build.sh
cd ../..

# Link package to virtual environment 
link tools/sortmerna-2.1/indexdb_rna
link tools/sortmerna-2.1/sortmerna

# Check if SortMeRNA was installed properly
indexdb_rna -h
sortmerna --version
```
<br>
Next we need to generate an index for the SortMeRNA database.
```{bash, eval = F}
# Set variable for location of index files
export sortmernaDB="tools/sortmerna-2.1"

# Set location for database
sortmernaREF=${sortmernaDB}/rRNA_databases/silva-arc-16s-id95.fasta,${sortmernaDB}/index/silva-arc-16s-id95:\
${sortmernaDB}/rRNA_databases/silva-arc-23s-id98.fasta,${sortmernaDB}/index/silva-arc-23s-id98:\
${sortmernaDB}/rRNA_databases/silva-bac-16s-id90.fasta,${sortmernaDB}/index/silva-bac-16s-id95:\
${sortmernaDB}/rRNA_databases/silva-bac-23s-id98.fasta,${sortmernaDB}/index/silva-bac-23s-id98:\
${sortmernaDB}/rRNA_databases/silva-euk-18s-id95.fasta,${sortmernaDB}/index/silva-euk-18s-id95:\
${sortmernaDB}/rRNA_databases/silva-euk-28s-id98.fasta,${sortmernaDB}/index/silva-euk-28s-id98

# Generate indexs (will take ~5 minutes)
indexdb_rna --ref $sortmernaREF 
```

#### 2e. Install STAR-aligner
"Spliced Transcripts Alignment to a Reference"  
<br>
**Link:** https://github.com/alexdobin/STAR  
```{bash, eval = F}
# Download STAR to the 'tools' folder
wget -P tools/ https://github.com/alexdobin/STAR/archive/2.5.2b.zip

# Unzip and remove compressed file
unzip tools/2.5.2b.zip -d tools/
rm -rf tools/2.5.2b.zip

# Link package to virtual environment 
link tools/STAR-2.5.2b/bin/Linux_x86_64/STAR

# Check if STAR was installed properly
STAR --version
```

#### 2f. Install Subread
"Subread package: high-performance read alignment, quantification and mutation discovery"  
<br>
**Link:** http://subread.sourceforge.net/  
**Link:** http://bioinf.wehi.edu.au/featureCounts/  

```{bash, eval = F}
# Download Subread to the 'tools' folder
wget -P tools/ https://sourceforge.net/projects/subread/files/subread-1.5.1/subread-1.5.1-Linux-x86_64.tar.gz

# Unzip and remove compressed file
tar -zxvf tools/subread-1.5.1-Linux-x86_64.tar.gz -C tools/
rm -rf tools/subread-1.5.1-Linux-x86_64.tar.gz

# Link package to virtual environment 
link tools/subread-1.5.1-Linux-x86_64/bin/featureCounts

# Check if Subread (featureCounts) was installed properly
featureCounts -v
```

#### 2g. Install MultiQC
"MultiQC searches a given directory for analysis logs and compiles a HTML report. It's a general use tool, perfect for summarising the output from numerous bioinformatics tools."  
**Link:** http://multiqc.info/  
```{bash, eval = F}
# Install MultiQC to the virtualenv
pip install multiqc
pip install click

# Check if MultiQC was installed properly
multiqc --version
```



### 3. Download host genome and annotation file
A host genome and annotation file are required for a complete RNA seq analysis. The annotation file contains the gene information associated with the coordinates of an alignment. The two files need to places in their respective folders before running the workflow. Additionally, There are different releases of the full mouse genome, so be aware which genome and which annotation file you are using for the sequencing alignment.  


#### 3a. Mouse genome
**Link:** http://www.gencodegenes.org/mouse_releases/current.html   

```{bash, eval = F}
# Download mouse genome to the 'genome' folder
wget -P genome/ ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M10/GRCm38.p4.genome.fa.gz

# Download genome annotation file to the 'annotation' folder
wget -P annotation/ ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M10/gencode.vM10.annotation.gtf.gz

# Decompress
gunzip genome/GRCm38.p4.genome.fa.gz
gunzip annotation/gencode.vM10.annotation.gtf.gz
```

#### 3b. Human genome
**Link:** http://www.gencodegenes.org/releases/current.html  

```{bash, eval = F}
# Download human genome to the 'genome' folder
wget -p genome/ ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_25/GRCh38.p7.genome.fa.gz

# Download genome annotation file to the 'annotation' folder
wget -P annotation/ ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_25/gencode.v25.annotation.gtf.gz

# Decompress
gunzip genome/GRCh38.p7.genome.fa.gz
gunzip annotation/gencode.v25.annotation.gtf.gz
```

#### 3c. Other genome
If you are not working with a mouse or human genome, see the following websites for a complete listing.  
https://genome.ucsc.edu/cgi-bin/hgGateway  
http://useast.ensembl.org/info/data/ftp/index.html  


### 4. Generate STAR aligner index
The STAR aligner requires an index to be created before aligning any ```fastq``` sequences. The index command requires the host genome of interest and associated annotation file as inputs. The command is best run on the cluster, but should be submitted as a job. Download the shell file and submit it as a job. Additionally you must specific the length of the sequences in the RNAseq run in the command. This is typically 50 or 101, but will depend on what was requested.

```{bash, eval = F}
# Download index job file
wget https://raw.githubusercontent.com/twbattaglia/RNAseq-workflow/master/make_index.sh

# Submit the make_index.sh as a job. Set the read length to 50bp.
qsub make_index.sh 50
```


### 5. Download workflow and change parameters
Once the index has been generated and all of the tools are installed, it is time to start the alignment workflow. The workflow is stored as a shell (.sh) file which submits an array job which will iterate over each input ```.fastq.gz``` file.
<br>  
Before submitting the command, make sure you change the variables within the file to reflect your data set.    
**qualityCutoff** : The minimum Phred score to retain nucleotides  [Default = 20]  
**trimLength** :  The minimum length of the sequence to keep after quality trimming. Typically 60% of the total length or greater. [Default = 20]  
```{bash, eval = F}
wget https://raw.githubusercontent.com/twbattaglia/RNAseq-workflow/master/run_workflow.sh
wget https://raw.githubusercontent.com/twbattaglia/RNAseq-workflow/master/postprocessing.sh
```



### 6a. Run the workflow in parallel
The job can be submitted to the cluster once you have changed the parameters to reflect your data and you have placed the ```.fastq.gz``` files into the folder, ```input```. The command is written to submit an array job, which means you must specify how many times the job should run. In this case it is dependent upon how many individual ```.fastq.gz``` files that are located in the ```input``` folder. If there are 6 different samples within the folder you would run the command ```qsub -t 1:6 run_workflow.sh```, but if you have 10 files, you would run the command ```qsub -t 1:10 run_workflow.sh```. The ```-t``` flag corresponds to how many different commands should be submitted.
```{bash, eval = F}
qsub -t 1:6 run_workflow.sh
```
**Note:** The command is written for SGE clusters, but if you have a different cluster set up, you can change the variable ```$SGE_TASK_ID``` to the correct array style variable.


### 6b. Process the output files and generate a report 
After the main workflow has finished running, you will have generated aligned ```.bam``` files for each sample. These files only give the coordinates for the sequence alignments on the genome, and must be summarized to give gene counts per sample, before any statistical analysis can be performed. To do this we perform further processing the output files using the ```post-processing.sh``` script There is no need for a parallel array script and the command can be submitted as is as long as no folder names/locations were not changed in main workflow step.  
```{bash, eval = F}
qsub post-processing.sh
```

### 6c. Outputs from running the workflow
Many files are generated after running the main workflow and the post-processing scripts. Below is a breakdown of each of the folders and their respective description of their outputs. Each step reports a log file for every sample which can be used to calculcated the number of sequences.  
```{bash, eval = F}
── output/
│   └── 1_initial_qc/ - FastQC quality reports for each sample
│   └── 2_trimmed_output/ - Trimmed reports for each sample
│   └── 3_rRNA
│       ├── aligned/ - Sequences aligned to rRNA databases
│       ├── filtered/ - Sequences with rRNA sequences removed
│       ├── logs/ - Logs from running SortMeRNA
│   └── 4_aligned_sequences
│       ├── aligned_bam/ - Main alignment files for each sample
│       ├── aligned_logs/ - Log from running STAR alignment step
│       ├── aligned_counts/ - STAR alignment counts output (for comparison with featureCounts)
│   └── 5_final_counts - Final gene count summarization output from running featureCounts
│   └── 6_multiQC - Overall workflow report
```




### 7. Differential expression analysis using R
Once the workflow has completed, you can now use the gene count table as an input into DESeq2 for statistical analysis using the R-programming language. It is highly reccomended to use **RStudio** when writing R code and generating R-related analyses. You can download RStudio for your system here: https://www.rstudio.com/products/rstudio/download/  
  
#### 7a. Generate a metadata file
One additional required file which is needed is a type of metadata file. This can be created in R or it can be imported as a text file. The mapping file must have sample identifiers that match the the resulting table with more columns that describe the sample (e.g Treatment).



#### 7b. Install required R-libraries
Multiple libraries are required to perform analysis, plotting and enrichment. 
```{R, eval = F}
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
biocLite("ggplot2")
biocLite("gplots")
biocLite("clusterProfiler")
biocLite("biomaRt")
biocLite("ReactomePA")
biocLite("DOSE")
biocLite("KEGG.db")
biocLite("pathview")
biocLite("org.Mm.eg.db")
biocLite("org.Hs.eg.db")
biocLite("pheatmap")
biocLite("genefilter")
biocLite("fdrtool")
biocLite("RColorBrewer")
biocLite("topGO")
biocLite("dplyr")
biocLite("NMF")
```


#### 7c. Using featureCounts output
One you have an R environment appropriatley set up, you can begin to import the featureCounts output of genes counts, found within the ```5_final_counts``` folder. This tutorial will use DESeq2 to normalize and perform the statistical analysis between sample groups. Be sure to copy the ```final_counts.txt``` file generate from featureCounts step to your set working directory, or specify the full location when importing the table. 
<br>  
If you would like to use an example featureCounts output, download the file here: (TODO). 


```{R, eval = F}
# Author 
# Date
# Summary of analysis


# - - - - - - - - - - - - - - -
# Load required libraries and
# set up environment
# - - - - - - - - - - - - - - -
library(DESeq2)
library(ggplot2)
library(fdrtool)
library(biomaRt)
library(pheatmap)
library(RColorBrewer)
library(clusterProfiler)

# Set working directory. (if needed)
setwd("~/path/to/working/directory/")


# - - - - - - - - - - - - - - -
# Import featureCounts data
# - - - - - - - - - - - - - - -

# Import gene counts table
# - skip first row (general command info)
# - make row names the gene identifiers
countdata <- read.table("final_counts.txt", header = TRUE, skip = 1, row.names = 1)

# Remove .bam + '..' from column identifiers
colnames(countdata) <- gsub(".bam", "", colnames(countdata), fixed = T)
colnames(countdata) <- gsub("..", "", colnames(countdata), fixed = T)

# Remove length/char columns
countdata <- countdata[ ,c(-1:-5)]
colnames(countdata) # make sure ID's are correct


# - - - - - - - - - - - - - - -
# Import metadata 
# (or create a new dataframe)
# - - - - - - - - - - - - - - -

# Import metadata file
# - make row names the matching sampleID's from the countdata
metadata <- read.delim("path/to/sample_mapping_file.txt", row.names = 1)

# Add sampleID's to the mapping file
metadata$sampleID <- row.names(metadata)

# Reorder sampleID's to match featureCounts column order. 
metadata <- metadata[match(colnames(countdata), metadata$sampleID),]


# - - - - - - - - - - - - - - - - - - - - - -
# Make DESeq2 object from featureCounts object 
# - - - - - - - - - - - - - - - - - - - - - -
# - countData : count dataframe
# - colData : sample metadata in the dataframe with row names as sampleID's
# - design : The design of the comparisons to use. 
#            Use (~) before the name of the column variable to compare
ddsMat <- DESeqDataSetFromMatrix(countData = counts,
                                 colData = metadata,
                                 design = ~Group)

# - - - - - - - - - - - - - - - - - - - - - -
# Find differential expressed genes
# - - - - - - - - - - - - - - - - - - - - - -
# Run DESEq2
ddsMat <- DESeq(ddsMat)

# Get results from testing with FDR adjust pvalues
res <- results(ddsMat, pAdjustMethod = "fdr", alpha = 0.05)

# Generate summary of testing. 
summary(res)

# Find the directionality of the fold change 
mcols(res, use.names = T)

# Order results by padj value 
res <- res[order(res_out$padj), ]
```


### 8. Gene annotation

#### 8a. Gathering gene annotation information
```{R, eval = F}
library(AnnotationDbi)
library(org.Mm.eg.db)

# Add gene full name
res$description <- mapIds(x = org.Mm.eg.db,
                          keys = row.names(res),
                          column = "GENENAME",
                          keytype = "ENSEMBL",
                          multiVals = "first")

# Add gene symbol
res$symbol <- mapIds(x = org.Mm.eg.db,
                     keys = row.names(res),
                     column = "SYMBOL",
                     keytype = "ENSEMBL",
                     multiVals = "first")

# Add ENTREZ ID
res$entrez <- mapIds(x = org.Mm.eg.db,
                     keys = row.names(res),
                     column = "ENTREZID",
                     keytype = "ENSEMBL",
                     multiVals = "first")

# Subset for only significant genes (q < 0.05)
res_sig <- subset(res, padj < 0.05)
summary(ddPups_res_sig)

```

#### 8b. Write all the important results to .txt files
```{R, eval = F}
# Write normalized gene counts to a .txt file
write.table(x = as.data.frame(counts(dds), normalized = T), 
            file = 'DESeq2_normalized_counts.txt', 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant normalized gene counts to a .txt file
write.table(x = counts(ddsMat[row.names(res_sig)], normalized = T), 
            file = 'DESeq2_normalized_counts_significant.txt', 
            sep = '\t', 
            quote = F, 
            col.names = NA)
            
# Write the annotated results table to a .txt file
write.table(x = as.data.frame(res), 
            file = "DESEq2_results_gene_annotated.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)

# Write significant annotated results table to a .txt file
write.table(x = as.data.frame(res_sig), 
            file = "DESEq2_results_gene_annotated_significant.txt", 
            sep = '\t', 
            quote = F,
            col.names = NA)
```


### 9. Plotting gene data

#### 9a. PCA of highest variance genes

```{R, eval = F}

# Convert all samples to rlog
ddsMat_rlog <- rlog(ddsMat, blind = FALSE)

# Plot PCA by column variable
plotPCA(ddsMat_rlog, intgroup = "Group")
```


#### 9b. Heatmap of significant genes

```{R, eval = F}
# Load libraries
library(pheatmap)
library(RColorBrewer)

# rlog transformation of the raw count data. 
# Necessary to show data in plots
rld <- rlog(ddsMat)

# Gather top genes and make matrix
mat <- assay(rld[row.names(res_out_sig)])

# Make colored column annotations.
# Choose which column variables you want to annotate the columns by.
annotation_col = data.frame(
  Treatment = factor(colData(rld)$Group),
  row.names = colData(rld)$SampleID
)

# Make Heatmap with pheatmap function.
# See more in documentation for customization
pheatmap(mat = mat, 
         color = colorRampPalette(brewer.pal(9, "YlOrBr"))(255), 
         scale = "row", 
         annotation_col = annotation_col,
         fontsize_row = 8)
```


#### 9c. Volcano plot
```{r, eval = F}

```

#### 9d. MA plot
```{r, eval = F}
plotMA(res, ylim=c(-5,5))

```

#### 9e. Plotting a single gene
```{r, eval = F}

plotCounts(dds, gene = "", intgroup = "dex")


```

### 10. Pathway analysis on differentially expressed genes. 
The code below assumes you have RNA seq results from the mouse genome.  
Get more information about clusterProfiler here: http://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html
```{R, eval = F}
# Load required libraries
library(clusterProfiler)
library(ReactomePA)
library(KEGG.db)
library(DOSE)
library(pathview)
library(org.Mm.eg.db)

# Create a matrix of gene entrez ID's and log fold changes
gene_matrix <- res_out_sig$logFC
names(gene_matrix) <- res_out_sig$entrez


# - - - - - - - - - - - - - 
# Enrich with KEGG database
# - - - - - - - - - - - - - 
kegg_enrich <- enrichKEGG(gene = names(gene_matrix),
                 organism = 'mouse',
                 pvalueCutoff = 0.05, 
                 readable = TRUE)
                 
# Get table of results
summary(kegg_enrich)

# Plot results
barplot(kegg_enrich, drop=TRUE, showCategory=12)

# Plot specific KEGG pathways (w fold change) with pathview
# pathway.id : KEGG pathway
pathview(gene.data = gene_matrix, 
         pathway.id = "04940", 
         species = "mouse", 
         map.symbol = T)


# - - - - - - - - - - - - - 
# Enrich with GO databse
# - - - - - - - - - - - - - 
go_enrich <- groupGO(gene = gene_matrix,
               organism = 'mouse',
               ont = "BP",
               level = 3,
               readable = TRUE)

# Get table of results
summary(go_enrich)

# Plot results
barplot(go_enrich, drop=TRUE, showCategory=12)


# - - - - - - - - - - - - - 
# Enrich with ReactomeDB
# - - - - - - - - - - - - - 
reactome_enrich <- enrichPathway(gene = gene_matrix, organism = 'mouse', pvalueCutoff = 0.05)
                 
# Get table of results
summary(reactome_enrich)
```

### 11. Going further with RNAseq analysis

You can the links below for a more in depth walk through of RNAseq analysis using R: 
* http://www.bioconductor.org/help/workflows/rnaseqGene/  
* http://bioconnector.org/workshops/r-rnaseq-airway.html
* http://www-huber.embl.de/users/klaus/Teaching/DESeq2Predoc2014.html
* http://www-huber.embl.de/users/klaus/Teaching/DESeq2.pdf
* https://web.stanford.edu/class/bios221/labs/rnaseq/lab_4_rnaseq.html
* http://www.rna-seqblog.com/which-method-should-you-use-for-normalization-of-rna-seq-data/




### Citations:

1. Andrews S. (2010). FastQC: a quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc  

2. Martin, Marcel. Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal, [S.l.], v. 17, n. 1, p. pp. 10-12, may. 2011. ISSN 2226-6089. Available at: <http://journal.embnet.org/index.php/embnetjournal/article/view/200>. doi:http://dx.doi.org/10.14806/ej.17.1.200.  

3. Kopylova E., Noé L. and Touzet H., "SortMeRNA: Fast and accurate filtering of ribosomal RNAs in metatranscriptomic data", Bioinformatics (2012), doi: 10.1093/bioinformatics/bts611

4. Dobin A, Davis CA, Schlesinger F, et al. STAR: ultrafast universal RNA-seq aligner. Bioinformatics. 2013;29(1):15-21. doi:10.1093/bioinformatics/bts635.

5. Lassmann et al. (2010) "SAMStat: monitoring biases in next generation sequencing data." Bioinformatics doi:10.1093/bioinformatics/btq614 [PMID: 21088025] 

6. Liao Y, Smyth GK and Shi W (2014). featureCounts: an efficient general purpose program for assigning sequence reads to genomic features. Bioinformatics, 30(7):923-30. 

7. Love MI, Huber W and Anders S (2014). “Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.” Genome Biology, 15, pp. 550.

8. Yu G, Wang L, Han Y and He Q (2012). “clusterProfiler: an R package for comparing biological themes among gene clusters.” OMICS: A Journal of Integrative Biology, 16(5), pp. 284-287.  

9. Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller. "MultiQC: Summarize analysis results for multiple tools and samples in a single report" Bioinformatics (2016). doi: 10.1093/bioinformatics/btw354. PMID: 27312411

